File: documentation-examples-scala/src/scala/dev/typr/foundations/docs/core/TemplateFrom.scala
package dev.typr.foundations.docs.core
import dev.typr.foundationssc.*
import dev.typr.foundationssc.data.*


@SuppressWarnings(Array("unused"))
object TemplateFrom:
  var tx: Transactor = null // placeholder

  //start
  // A case class gives names to each template parameter
  case class InsertUser(name: String, email: String)

  // .from() maps case class fields to template params
  val insertUser: Template.From[InsertUser, Int] =
    Fragment.of("INSERT INTO users(name, email) VALUES(")
      .param(PgTypes.text)
      .append(", ")
      .param(PgTypes.text)
      .append(")")
      .update()
      .from(_.name, _.email)

  // Callers pass the case class
  def createUser(): Int =
    insertUser.on(InsertUser("Alice", "alice@example.com"))
      .transact(tx)
  //stop

File: documentation-examples-java/src/java/dev/typr/foundations/docs/core/TemplateFrom.java
package dev.typr.foundations.docs.core;

import dev.typr.foundations.Fragment;
import dev.typr.foundations.PgTypes;
import dev.typr.foundations.Template;
import dev.typr.foundations.Transactor;

@SuppressWarnings("unused")
public class TemplateFrom {
    Transactor tx = null; // placeholder

    //start
    // A record gives names to each template parameter
    record InsertUser(String name, String email) {}

    // .from() maps record fields to template params
    Template.From<InsertUser, Integer> insertUser =
        Fragment.of("INSERT INTO users(name, email) VALUES(")
            .param(PgTypes.text)
            .append(", ")
            .param(PgTypes.text)
            .append(")")
            .update()
            .from(InsertUser::name, InsertUser::email);

    // Callers pass the record
    int createUser() {
        return insertUser.on(new InsertUser("Alice", "alice@example.com"))
            .transact(tx);
    }
    //stop
}
